### PR TITLE
Future class

### PR DESCRIPTION
**Pull request by Hroy02**

### CODE REVIEW

File: src/main/java/bgu/spl/mics/Future.java [hunk 1]
```diff
@@ -29,12 +32,15 @@ public Future() {
...
      *
      * @return return the result of type T if it is available, if not wait until it is available.
-     *
      * @post (isDone = = true)
      */
-    public T get() {
-        //TODO: implement this.
-        return null;
+    public synchronized T get() {
+        while (_done == false)
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/Future.java]: I prefer !_done (:

File: src/main/java/bgu/spl/mics/Future.java [hunk 2]
```diff
@@ -43,19 +49,20 @@ public T get() {
      * @pre (isDone () == false)
      * @post (isDone () == true && get() == result)
      */
-    public void resolve(T result) {
-        //TODO: implement this.
+    public synchronized void resolve(T result) {
+        if (_done) return;
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/Future.java]: I believe we can asume it won't happen.

File: src/main/java/bgu/spl/mics/Future.java [hunk 3]
```diff
@@ -43,19 +49,20 @@ public T get() {
      * @pre (isDone () == false)
      * @post (isDone () == true && get() == result)
      */
-    public void resolve(T result) {
-        //TODO: implement this.
+    public synchronized void resolve(T result) {
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/Future.java:52]: why do you need to synchronize this method?

File: src/main/java/bgu/spl/mics/Future.java [hunk 4]
```diff
@@ -65,15 +72,17 @@ public boolean isDone() {
...
-     *
      * @return return the result of type T if it is available, if not, wait for {@code timeout} TimeUnits {@code unit}.
      * If time has elapsed, return null.
-     *
      * @post (isDone () == true || current_time >= start_time + timeout
      */
     public T get(long timeout, TimeUnit unit) {
-        //TODO: implement this.
-        return null;
+        while (_done == false)
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/Future.java]: same

File: src/main/java/bgu/spl/mics/Future.java (hunk 1 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/Future.java]: Done

File: src/main/java/bgu/spl/mics/Future.java (hunk 2 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/Future.java]: Right, it's synced

File: src/main/java/bgu/spl/mics/Future.java (hunk 3 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/Future.java:52]: Because we want to be able to use wait and notify

Comment (IdanTo1) [reply src/main/java/bgu/spl/mics/Future.java:52]: I think you can use notify without synchronized

**Review by IdanTo1 — APPROVED** [review]

File: src/main/java/bgu/spl/mics/Future.java (hunk 3 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/Future.java:52]: There are multiple reasons, some are explained [here](https://stackoverflow.com/questions/2779484/why-must-wait-always-be-in-synchronized-block)

Comment (IdanTo1) [reply src/main/java/bgu/spl/mics/Future.java:52]: Yes but they are talking about wait, not notify..

Comment (Hroy02) [reply src/main/java/bgu/spl/mics/Future.java:52]: explained [here](https://stackoverflow.com/questions/8817266/why-should-the-notify-method-be-inside-a-synchronized-block)

**Review by IdanTo1 — APPROVED** [review]

**Pull request merged**